inet/vibe/inet/url.d

```diff
@@ -19,11 +19,75 @@ import std.exception;
 import std.string;
 import std.traits : isInstanceOf;
 import std.ascii : isAlpha, isASCII, toLower;
-import std.uri: encode;
+import std.uri: decode, encode;
 
 import core.checkedint : addu;
 
 
+/** Parses a user-entered URL with relaxed parsing rules.
+
+	Unlike `URL.parse`, this allows the URL to use special charaters as part of
+	the host name and path, automatically emplying punycode or percent-encoding
+	to convert this to a valid URL.
+
+	Also, relative URL references are supported by supplying a base URL.
+
+	Params:
+		url = String representation of the URL
+		base_url = Base
+*/
+URL parseUserURL(string url, URL base_url = URL.init)
+{
+	// test if this is a URL or a relative (path) reference
+	auto sidx = url.indexOf(':');
+	if (sidx < 0) {
+		// TODO: handle query strings and anchors
+
+		if (url.representation.canFind('%')) {
+			try url.decode();
+			catch (Exception e) url = url.encode();
+		} else url = url.encode();
+
+		auto path = InetPath(url);
+
+		if (base_url != URL.init) {
+			if (path.absolute)
+				base_url.path = path;
+			else if (base_url.path.endsWithSlash)
+				base_url ~= path;
+			else base_url.path = base_url.path.parentPath ~ path;
+			return base_url;
```

Comment:
Personally I have different expectations for the "base URL". I assumed that if the user provides a "full" URL, then it is used. Otherwise, I provide a default. I am not saying my expectations are more sensible than what this function is doing, but it would be best if what the function does was documented a bit more.

Something like:
> If a `base` URL is provided, it will always be used, and only the path part of `url` will be taken into account. Moreover, if the base URL is not a directory (e.g. `http://example.com/foo`), the file path (here: `foo`) will be replaced. To ensure the base URL is appended to, make sure it ends with a directory (e.g. `http://example.com/foo/`).

Reply:
Having a few documented unittests serving as example would help, so instead of having examples between parenthesis, one could write:
```D
unittest
{
    // More basic tests...
    
    // When a base URL is provided, it is always used
    assert(parseUserURL("vibe-d.org/foo/bar", URL("http://example.com")) == URL("http://example.com/foo/bar"));
    // Etc...
}
```

Actually while writing the example I realized one thing: The base URL is not *always* used. If the user-inputted string is a full URL with proto, then it is ignored, another thing that might be surprising to users.

Reply:
This is a good point, right now the logic is to treat something that is not explicitly a URL (with ":") as a path, whereas the expectation for entering something in the browser's address bar is to treat a non-absolute path as a http(s) URL. I'll probably just remove the whole `base_url` logic then, because I didn't really intend to invest time into the relaxed parsing at this point.

But fortunately we can simply follow the W3C guidelines here.

Reply:
I think what the default protocol should be is up to the application.
A browser will try `https`, then `http`, but if the URL is read from the configuration file of an FTP client, then that would be the default protocol (or one of the alternative).
In our case, we had a P2P protocol, so we didn't want people to have to type `agora://host.foo`.
IOW, instead of a base URL, maybe consider a default (as in, used in absence of explicit value) protocol ?

Reply:
I have already removed the relative reference handling. I was thinking about the way relative references can be used within e.g. HTML documents, but thinking about it, I can't really see the use case from a user's point of view. But having a default protocol parameter makes a lot of sense.